packages/react-keytips/src/components/Keytips/useKeytips.tsx

```diff
@@ -0,0 +1,274 @@
+        if (currentKeytip.target) {
+          currentKeytip?.onReturn?.(ev, {
+            event: ev,
+            type: 'keydown',
+            targetElement: currentKeytip.target,
+          });
+        }
+      }
+      currentSequence.current = '';
+      tree.getBack();
+      showKeytips(tree.getChildren());
+      if (tree.currentKeytip.current === undefined) {
+        setInKeytipMode(false);
+      }
+    },
+    [inKeytipMode]
+  );
+
+  useHotkeys([
+    [startSequence, handleEnterKeytipMode],
+    [returnSequence, handleReturnSequence],
+    [exitSequence, handleExitKeytipMode],
+    ['tab', handleExitKeytipMode],
+    ['enter', handleExitKeytipMode],
+    ['space', handleExitKeytipMode],
+  ]);
+
+  React.useEffect(() => {
+    const onKeytipAdded = (keytip: KeytipWithId) => {
+      tree.addNode(keytip);
+
+      setKeytips((prev) => ({
+        ...prev,
+        [keytip.uniqueId]: {
+          ...keytip,
+          id: sequencesToID(keytip.keySequences),
+        },
+      }));
+
+      if (tree.isCurrentKeytipParent(keytip)) {
+        showKeytips(tree.getChildren());
+      }
+    };
+
+    const onKeytipRemoved = (keytip: KeytipWithId) => {
+      tree.removeNode(keytip);
+      setKeytips((prev) => {
+        const newKeytips = { ...prev };
+        delete newKeytips[keytip.uniqueId];
+        return newKeytips;
+      });
+    };
+
+    const onKeytipUpdated = (keytip: KeytipWithId) => {
+      tree.updateNode(keytip);
+      setKeytips((prev) => ({
+        ...prev,
+        [keytip.uniqueId]: {
+          ...keytip,
+          id: sequencesToID(keytip.keySequences),
+        },
+      }));
+    };
+
+    subscribe(EVENTS.KEYTIP_ADDED, onKeytipAdded);
+    subscribe(EVENTS.KEYTIP_UPDATED, onKeytipUpdated);
+    subscribe(EVENTS.KEYTIP_REMOVED, onKeytipRemoved);
+    return () => {
+      unsubscribe(EVENTS.KEYTIP_ADDED, onKeytipAdded);
+      unsubscribe(EVENTS.KEYTIP_UPDATED, onKeytipUpdated);
+      unsubscribe(EVENTS.KEYTIP_REMOVED, onKeytipRemoved);
+    };
+  }, []);
+
+  React.useEffect(() => {
+    const handleDismiss = (ev: Event) => {
+      if (inKeytipMode) {
+        handleExitKeytipMode(ev as KeyboardEvent);
+      }
+    };
+
+    const addEventListeners = () => {
+      targetDocument?.addEventListener('mouseup', handleDismiss);
+      targetDocument?.defaultView?.addEventListener('resize', handleDismiss);
+      targetDocument?.defaultView?.addEventListener('scroll', handleDismiss);
+    };
+
+    const removeEventListeners = () => {
+      targetDocument?.removeEventListener('mouseup', handleDismiss);
+      targetDocument?.defaultView?.removeEventListener('resize', handleDismiss);
+      targetDocument?.defaultView?.removeEventListener('scroll', handleDismiss);
+    };
+
+    addEventListeners();
+    return removeEventListeners;
+  }, [targetDocument, inKeytipMode, handleExitKeytipMode]);
+
+  React.useEffect(() => {
+    const handleKeyDown = (ev: KeyboardEvent) => {
```

Comment:
The complexity of this method is quite high; it might be beneficial to break it down into smaller, named functions.